### PR TITLE
add initial version of encoder and sync

### PR DESCRIPTION
**Pull request by nityanandagohain**
PR for adding zap otlp sync and encoder

Things not present in the PR
* Support for namespaces
* Zap array fields
* Zap object fields
* Zap complex fields

### CODE REVIEW

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	enc.log.Attributes = append(enc.log.Attributes, &v1.KeyValue{
+		Key:   key,
+		Value: val,
+	})
+}
+
+func addFields(enc zapcore.ObjectEncoder, fields []zapcore.Field) {
+	for i := range fields {
+		fields[i].AddTo(enc)
+	}
```
Comment (srikanthccv) [inline zap_otlp_encoder/zap_otlp_encoder.go:254]: How are these fields retrieved back? I see syncer is just unmarshalling to logrecord which won't contain the fields?

File: zap_otlp_sync/zap_otlp_sync.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+	return nil
+}
+
+func (l *OtelSyncer) Write(record []byte) (n int, err error) {
+	// create deep copyy of the data
+	data := make([]byte, len(record))
+	copy(data, record)
+	l.values = append(l.values, data)
+	if len(l.values) >= l.batchSize {
```
Comment (srikanthccv) [inline zap_otlp_sync/zap_otlp_sync.go:114]: Look at this https://github.com/open-telemetry/opentelemetry-go/blob/4d473d105a5f59a58d153e7c492a4ea0ee027871/sdk/trace/batch_span_processor.go#L287-L360

File: zap_otlp_sync/zap_otlp_sync.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+func (l OtelSyncer) Sync() error {
+	if err := l.pushToSigNoz(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (l OtelSyncer) Close() error {
+	l.close()
```
Comment (srikanthccv) [inline zap_otlp_sync/zap_otlp_sync.go:132]: This should flush any remaining items

File: zap_otlp_encoder/zap_otlp_encoder.go (hunk 1 above)
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:254]: What do you mean by retrieved back ? do you mean how it is sent to otel collector ?

File: zap_otlp_sync/zap_otlp_sync.go (hunk 3 above)
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go:132]: Makes sense.

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+type otlpEncoder struct {
+	*zapcore.EncoderConfig
+	buf            *buffer.Buffer
+	spaced         bool
+	openNamespaces int
+
+	log *lpb.LogRecord
+
+	// for encoding generic values by reflection
+	reflectBuf *buffer.Buffer
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:53]: **buf**  :-  this is the buffer that zap expects from an encoder.
**log**  :- since we want to store in in otel log format instead of manually adding all the data to buffer, I am adding it to the **log** and then finally converting the log to buffer which zap expects

File: zap_otlp_encoder/zap_otlp_encoder.go (hunk 1 above)
Comment (srikanthccv) [reply zap_otlp_encoder/zap_otlp_encoder.go:254]: Yes, where are they deserialised?

Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:254]: It's done by the pushData function in the zap_otlp_sync.

File: zap_otlp_sync/zap_otlp_sync.go (hunk 2 above)
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go]: Thanks I have made the changes according to the above implementation.

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+	reflectEnc zapcore.ReflectedEncoder
+}
+
+// NewkvEncoder creates a key=value encoder
+func NewOTLPEncoder(cfg zapcore.EncoderConfig) zapcore.Encoder {
+	if cfg.SkipLineEnding {
+		cfg.LineEnding = ""
+	} else if cfg.LineEnding == "" {
+		cfg.LineEnding = zapcore.DefaultLineEnding
+	}
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go]: Ahh this is not required for us.

File: zap_otlp_encoder/zap_otlp_encoder.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go]: Not using the reflectedBuf for now so have commented it out.

I am fully not clear on why it is required in the json encoder present here https://github.com/uber-go/zap/blob/382e2511e51cda8afde24f9e6e741f934308edfa/zapcore/json_encoder.go#L61

File: zap_otlp_sync/zap_otlp_sync.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+		options.ResourceSchema,
+		instrumentationScope,
+		[][]byte{},
+		make(chan []byte, options.BatchSize), // keeping queue size as batch size
+		collpb.NewLogsServiceClient(conn),
+		options.BatchSize,
+		make(chan bool),
+		make(chan bool),
+		sync.Mutex{},
+	}
```
Comment (srikanthccv) [inline zap_otlp_sync/zap_otlp_sync.go:90]: Can you add the `{Name: Value}` format here? It's hard to follow what each of these corresponds to.

File: zap_otlp_sync/zap_otlp_sync.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func (l *OtelSyncer) Close() error {
+	err := l.Sync()
+	if err != nil {
+		return err
+	}
+	l.closeExporter <- true
+	l.close()
+	return nil
```
Comment (srikanthccv) [inline zap_otlp_sync/zap_otlp_sync.go:183]: This should wait for the last flush to finish before returning.

File: zap_otlp_encoder/zap_otlp_encoder.go (hunk 1 above)
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:248]: Sure let me write a couple of tests.

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+func (enc *otlpEncoder) AddInt8(k string, v int8)       {}
+func (enc *otlpEncoder) AddUint(k string, v uint)       {}
+func (enc *otlpEncoder) AddUint32(k string, v uint32)   {}
+func (enc *otlpEncoder) AddUint16(k string, v uint16)   {}
+func (enc *otlpEncoder) AddUint8(k string, v uint8)     {}
+func (enc *otlpEncoder) AddUintptr(k string, v uintptr) {}
+
+func (enc *otlpEncoder) Clone() zapcore.Encoder {
+	clone := enc.clone()
+	clone.buf.Write(enc.buf.Bytes())
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:152]: Yeah, since we are mostly working with the log object it should be added.

File: zap_otlp_sync/zap_otlp_sync.go (hunk 6 above)
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go:90]: Sure

File: zap_otlp_sync/zap_otlp_sync.go (hunk 7 above)
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go:183]: Makes sense

File: zap_otlp_encoder/zap_otlp_encoder.go (hunk 1 above)
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:250]: So fields are moved to attributes, if it's something like top-level key of the OTEL log model like traceId, then it is added respectively.

I have added some tests, please check once.

File: zap_otlp_sync/zap_otlp_sync.go (hunk 7 above)
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go:183]: Since all the signals  are handled by the same processQueue function, the closeExporter signal will be run only after sync is complete in line no Close() function.

File: zap_otlp_sync/zap_otlp_sync.go [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+		rattrs = &rv1.Resource{Attributes: attrs}
+	}
+
+	var instrumentationScope *cv1.InstrumentationScope
+	if options.Scope != nil {
+		instrumentationScope = &cv1.InstrumentationScope{
+			Name:    options.Scope.Name,
+			Version: options.Scope.Version,
+		}
+	}
```
Comment (srikanthccv) [inline zap_otlp_sync/zap_otlp_sync.go:75]: This should be removed in favour of using logger name for scope

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+	// reflectBuf *buffer.Buffer
+	// reflectEnc zapcore.ReflectedEncoder
+}
+
+// NewkvEncoder creates a key=value encoder
+func NewOTLPEncoder(cfg zapcore.EncoderConfig) zapcore.Encoder {
+	// If no EncoderConfig.NewReflectedEncoder is provided by the user, then use default
+	if cfg.NewReflectedEncoder == nil {
+		cfg.NewReflectedEncoder = defaultReflectedEncoder
+	}
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:62]: Sure.

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+
+func (enc *otlpEncoder) AddInt(k string, v int)         {}
+func (enc *otlpEncoder) AddInt32(k string, v int32)     {}
+func (enc *otlpEncoder) AddInt16(k string, v int16)     {}
+func (enc *otlpEncoder) AddInt8(k string, v int8)       {}
+func (enc *otlpEncoder) AddUint(k string, v uint)       {}
+func (enc *otlpEncoder) AddUint32(k string, v uint32)   {}
+func (enc *otlpEncoder) AddUint16(k string, v uint16)   {}
+func (enc *otlpEncoder) AddUint8(k string, v uint8)     {}
+func (enc *otlpEncoder) AddUintptr(k string, v uintptr) {}
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:148]: We need these function definitions according to the encoder interface, but since we are writing to log record directly these are empty.

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+		fields[i].AddTo(enc)
+	}
+}
+
+func defaultReflectedEncoder(w io.Writer) zapcore.ReflectedEncoder {
+	enc := json.NewEncoder(w)
+	// For consistency with our custom JSON encoder.
+	enc.SetEscapeHTML(false)
+	return enc
+}
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:258]: nope

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+
+var levelMap = map[string]uint{
+	"trace": 1,
+	"debug": 5,
+	"info":  9,
+	"warn":  13,
+	"error": 17,
+	"fatal": 21,
+}
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:168]: Have updated this to a more suitable one, but mapping of these texts are not available.

[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:168]: Instead of assigning numbers individually, have used it from the go.opentelemetry.io/proto/otlp/logs/v1

File: zap_otlp_sync/zap_otlp_sync.go (hunk 7 above)
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go:183]: Let me add that

Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go:187]: Have added a waitgroup.

File: zap_otlp_sync/zap_otlp_sync.go [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+	BatchSize          int
+	BatchIntervalInSec int
+	ResourceSchema     string
+	Scope              *instrumentation.Scope
+	Resource           *resource.Resource
+}
+
+func NewOtlpSyncer(conn *grpc.ClientConn, options Options) *OtelSyncer {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go]: Yeah, this was a bug, have updated the code to use it properly.

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+		final.log.TimeUnixNano = uint64(ent.Time.UnixNano())
+	}
+	if ent.LoggerName != "" && final.NameKey != "" {
+		final.log.Attributes = append(enc.log.Attributes, &v1.KeyValue{
+			Key: "logger",
+			Value: &v1.AnyValue{
+				Value: &v1.AnyValue_StringValue{
+					StringValue: ent.LoggerName,
+				},
+			},
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:184]: Makes sense 👍

File: zap_otlp_sync/zap_otlp_sync.go (hunk 9 above)
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go]: Have added this changes, to use a map for instrumentation scope using logger name

File: zap_otlp_encoder/zap_otlp_encoder.go [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+		}
+	}
+
+	addFields(final, fields)
+	data, err := proto.Marshal(final.log)
+	if err != nil {
+		return nil, err
+	}
+
+	final.buf.AppendString(logger + "#SIGNOZ#" + string(data))
```
Comment (srikanthccv) [inline zap_otlp_encoder/zap_otlp_encoder.go:204]: We can still keep the name as a part of the attributes here instead of this separator logic. And in the transporter/sync program just delete the attribute and map it to the instrumentation scope.

Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go:204]: Attribute is an array which will result in searching and removing the attribute, I feel this one is a simpler implementation compared to that, but happy to hear your comments.

File: zap_otlp_encoder/zap_otlp_encoder.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply zap_otlp_encoder/zap_otlp_encoder.go]: Ahh right, I have updated the functions and added tests for them

File: zap_otlp_sync/zap_otlp_sync.go [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+	}
+
+	if l.res != nil {
+		rl.SchemaUrl, rl.Resource = l.resSchema, l.res
+	}
+	_, err = l.client.Export(context.Background(), &collpb.ExportLogsServiceRequest{
+		ResourceLogs: []*lpb.ResourceLogs{rl},
+	})
+	// TODO: how to handle partial failure and error
+	if err != nil {
```
Comment (makeavish) [inline zap_otlp_sync/zap_otlp_sync.go:152]: Should we also log the error here?

Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go:152]: we can use a fmt.Errorf here I guess, but not totally sure.

Comment (makeavish) [reply zap_otlp_sync/zap_otlp_sync.go:152]: Let's log it, I have seen exporters logging errors.

File: zap_otlp_encoder/zap_otlp_encoder.go (hunk 16 above)
Comment (srikanthccv) [reply zap_otlp_encoder/zap_otlp_encoder.go:222]: Since we want to advocate this to a larger audience, I suggested the attributes approach, given that attributes is usually small-size array. And want to avoid buggy cases where separator keyword appears in `data` rare but still there is a chance. I don't have a strong opinion here.

File: zap_otlp_sync/zap_otlp_sync.go (hunk 17 above)
Comment (srikanthccv) [reply zap_otlp_sync/zap_otlp_sync.go:153]: Curious why `fmt`, not `log`?

**Review by srikanthccv — APPROVED** [review]
Thanks for addressing my comments and being patient for several rounds of reviews. No major comments. LGTM, you can iterate on this.

**Review by makeavish — APPROVED** [review]

File: zap_otlp_sync/zap_otlp_sync.go (hunk 17 above)
Comment (nityanandagohain) [reply zap_otlp_sync/zap_otlp_sync.go:153]: Yah log will be better, it will include ts as well. Let me update it.
